### DISCUSSION TITLE
External Kalman Filter

### DISCUSSION THREAD

**Discussion by CalvinKoung** [bitcraze/discussions#987] [Q&A]
Hi, how can I improve the EKF of the crazyflie to increase the flight stability in the UWB indoor navigation ? Is there have the files or documents to research ? thanks

**Comment by knmcguire** [comment]
Hi! I'll just give you a bunch of links to read up on to get you on track:

- Doc about the stabilizer module (including EKF): https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/sensor-to-control/
- Info about the loco deck mearurement model for the EKF. https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/loco-positioning-system/
- More in-depth information about the loco positioning system: https://www.bitcraze.io/documentation/repository/lps-node-firmware/master/
- Parameters related to the loco positioning system including standard deviation of the measurement https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#loco

This is all assuming that you have some experience with extended kalman filters or at least know the basics, but if not then it would be good to look some explanation or videos online.

**Reply by CalvinKoung** [reply]
ok, thanks

**Reply by knmcguire** [reply]
ah also i'd like to add the outlier filter that you can tune as well. This blogpost is about that: https://www.bitcraze.io/2023/03/new-loco-tdoa-outlier-filter/

**Reply by CalvinKoung** [reply]
ok, thanks your help.

**Reply by knmcguire** [reply]
Also, don't forget to answer other discussions open questions about this. There might be more things going on then only tuning your EKF. https://github.com/orgs/bitcraze/discussions/975

**Reply by CalvinKoung** [reply]
Sorry, I forgot to reply. And ten, Can I ask TDoA must use 8 anchors can't use 6 anchors in this page ?

**Reply by krichardsson** [reply]
In theory 4 anchors should be enough, but I think 6 is more reasonable.

**Reply by CalvinKoung** [reply]
ok, thanks.

**Comment by krichardsson** [comment]
I would also like to add that there has been (and is ongoing) research in this area. There is an implementation of a robust M-estimation-based Kalman filter in the firmware that you can use by setting the [`kalman.robustTdoa`](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/api/params/#kalmanrobusttdoa) parameter. You can [find documentation and a reference to the paper on our web](https://www.bitcraze.io/documentation/repository/crazyflie-firmware/master/functional-areas/loco-positioning-system/#m-estimation-based-robust-kalman-filter).
